File: python/rag/sqlrooms_rag/prepare/metadata.py
"""
Metadata creation and storage for embedding databases.
"""

import yaml
from datetime import datetime
from pathlib import Path
from typing import Dict, Any

import duckdb


def calculate_chunk_stats(nodes: list) -> Dict[str, Any]:
    """
    Calculate statistics about the chunks.
    
    Args:
        nodes: List of llama-index Node objects
    
    Returns:
        Dictionary with chunk statistics
    """
    if not nodes:
        return {
            'total_chunks': 0,
            'min_chunk_size': 0,
            'max_chunk_size': 0,
            'median_chunk_size': 0,
            'mean_chunk_size': 0,
            'total_characters': 0,
        }
    
    # Get chunk sizes
    chunk_sizes = [len(node.text) if hasattr(node, 'text') else 0 for node in nodes]
    chunk_sizes.sort()
    
    return {
        'total_chunks': len(nodes),
        'min_chunk_size': min(chunk_sizes),
        'max_chunk_size': max(chunk_sizes),
        'median_chunk_size': chunk_sizes[len(chunk_sizes) // 2],
        'mean_chunk_size': sum(chunk_sizes) // len(chunk_sizes),
        'total_characters': sum(chunk_sizes),
    }


def create_metadata(
    documents: list,
    nodes: list,
    embedding_provider: str,
    embed_model_name: str,
    embed_dim: int,
    chunk_size: int,
    use_markdown_chunking: bool,
    include_headers_in_chunks: bool,
    header_weight: int,
) -> Dict[str, Any]:
    """
    Create comprehensive metadata about the embedding preparation.
    
    This metadata is crucial for:
    - Ensuring query-time uses the same model
    - Debugging and reproducibility
    - Understanding the dataset
    - Version compatibility
    
    Args:
        documents: List of source documents
        nodes: List of chunk nodes
        embedding_provider: Provider used (huggingface/openai)
        embed_model_name: Name of embedding model
        embed_dim: Embedding dimension
        chunk_size: Configured chunk size
        use_markdown_chunking: Whether markdown-aware chunking was used
        include_headers_in_chunks: Whether headers were included
        header_weight: Header repetition weight
    
    Returns:
        Dictionary with comprehensive metadata
    """
    chunk_stats = calculate_chunk_stats(nodes)
    
    # Count unique source files
    unique_files = set()
    for doc in documents:
        if hasattr(doc, 'metadata') and doc.metadata:
            file_path = doc.metadata.get('file_path', '')
            if file_path:
                unique_files.add(file_path)
    
    metadata = {
        'version': '1.0',
        'created_at': datetime.utcnow().isoformat() + 'Z',
        'embedding': {
            'provider': embedding_provider,
            'model': embed_model_name,
            'dimensions': embed_dim,
        },
        'chunking': {
            'strategy': 'markdown-aware' if use_markdown_chunking else 'size-based',
            'chunk_size': chunk_size,
            'include_headers': include_headers_in_chunks,
            'header_weight': header_weight if include_headers_in_chunks else 0,
        },
        'source_documents': {
            'total_documents': len(documents),
            'unique_files': len(unique_files),
            'total_characters': sum(len(doc.text) for doc in documents if hasattr(doc, 'text')),
        },
        'chunks': chunk_stats,
        'capabilities': {
            'hybrid_search': True,
            'fts_enabled': True,
            'source_documents_stored': True,
        }
    }
    
    return metadata


def store_metadata_in_db(db_path: Path, metadata: Dict[str, Any], verbose: bool = True) -> None:
    """
    Store metadata in the DuckDB database for runtime access.
    
    Creates a 'metadata' table with key-value pairs that can be queried
    at runtime to validate model compatibility.
    
    Args:
        db_path: Full path to the DuckDB database file
        metadata: Metadata dictionary
        verbose: Whether to print progress messages
    """
    if verbose:
        print("Storing metadata in database...")
    
    conn = duckdb.connect(str(db_path))
    
    try:
        # Create metadata table
        conn.execute("""
            CREATE TABLE IF NOT EXISTS embedding_metadata (
                key VARCHAR PRIMARY KEY,
                value VARCHAR,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        
        # Flatten metadata for storage
        flat_metadata = {
            'version': metadata['version'],
            'created_at': metadata['created_at'],
            # Embedding info
            'embedding_provider': metadata['embedding']['provider'],
            'embedding_model': metadata['embedding']['model'],
            'embedding_dimensions': str(metadata['embedding']['dimensions']),
            # Chunking info
            'chunking_strategy': metadata['chunking']['strategy'],
            'chunk_size': str(metadata['chunking']['chunk_size']),
            'include_headers': str(metadata['chunking']['include_headers']),
            'header_weight': str(metadata['chunking']['header_weight']),
            # Document stats
            'total_source_documents': str(metadata['source_documents']['total_documents']),
            'unique_files': str(metadata['source_documents']['unique_files']),
            'source_total_characters': str(metadata['source_documents']['total_characters']),
            # Chunk stats
            'total_chunks': str(metadata['chunks']['total_chunks']),
            'min_chunk_size': str(metadata['chunks']['min_chunk_size']),
            'max_chunk_size': str(metadata['chunks']['max_chunk_size']),
            'median_chunk_size': str(metadata['chunks']['median_chunk_size']),
            'mean_chunk_size': str(metadata['chunks']['mean_chunk_size']),
            'chunks_total_characters': str(metadata['chunks']['total_characters']),
            # Capabilities
            'hybrid_search_enabled': str(metadata['capabilities']['hybrid_search']),
            'fts_enabled': str(metadata['capabilities']['fts_enabled']),
            'source_documents_stored': str(metadata['capabilities']['source_documents_stored']),
        }
        
        # Insert/update metadata
        for key, value in flat_metadata.items():
            conn.execute("""
                INSERT INTO embedding_metadata (key, value)
                VALUES (?, ?)
                ON CONFLICT (key) DO UPDATE SET
                    value = EXCLUDED.value
            """, [key, value])
        
        if verbose:
            print(f"✓ Stored {len(flat_metadata)} metadata entries in database")
            
    finally:
        conn.close()


def save_metadata_yaml(db_path: Path, metadata: Dict[str, Any], verbose: bool = True) -> None:
    """
    Save metadata as a YAML file next to the database.
    
    This provides human-readable documentation and makes it easy to
    check settings without opening the database.
    
    Args:
        db_path: Full path to the DuckDB database file
        metadata: Metadata dictionary
        verbose: Whether to print progress messages
    """
    # Determine YAML file path (same location as .duckdb, but .yaml extension)
    yaml_path = db_path.with_suffix('.yaml')
    
    if verbose:
        print(f"Saving metadata to {yaml_path.name}...")
    
    try:
        with open(yaml_path, 'w') as f:
            yaml.dump(metadata, f, default_flow_style=False, sort_keys=False)
        
        if verbose:
            print(f"✓ Metadata saved to {yaml_path}")
            
    except Exception as e:
        if verbose:
            print(f"Warning: Could not save YAML metadata: {e}")

File: packages/ai-rag/src/prepare/metadata.ts
/**
 * Metadata creation and storage for embedding databases.
 * Mirrors: python/rag/sqlrooms_rag/prepare/metadata.py
 */

import type {DuckDbConnector} from '@sqlrooms/duckdb';

export type ChunkStats = {
  totalChunks: number;
  minChunkSize: number;
  maxChunkSize: number;
  medianChunkSize: number;
  meanChunkSize: number;
  totalCharacters: number;
};

export type EmbeddingMetadata = {
  version: string;
  createdAt: string;
  embedding: {
    provider: string;
    model: string;
    dimensions: number;
  };
  chunking: {
    strategy: 'markdown-aware' | 'size-based';
    chunkSize: number;
    includeHeaders: boolean;
    headerWeight: number;
  };
  sourceDocuments: {
    totalDocuments: number;
    uniqueFiles: number;
    totalCharacters: number;
  };
  chunks: ChunkStats;
  capabilities: {
    hybridSearch: boolean;
    ftsEnabled: boolean;
    sourceDocumentsStored: boolean;
  };
};

/**
 * Calculate statistics about chunks.
 */
export function calculateChunkStats(chunks: Array<{text: string}>): ChunkStats {
  if (chunks.length === 0) {
    return {
      totalChunks: 0,
      minChunkSize: 0,
      maxChunkSize: 0,
      medianChunkSize: 0,
      meanChunkSize: 0,
      totalCharacters: 0,
    };
  }

  const sizes = chunks.map((c) => c.text.length).sort((a, b) => a - b);
  const total = sizes.reduce((a, b) => a + b, 0);

  return {
    totalChunks: sizes.length,
    minChunkSize: sizes[0] ?? 0,
    maxChunkSize: sizes[sizes.length - 1] ?? 0,
    medianChunkSize: sizes[Math.floor(sizes.length / 2)] ?? 0,
    meanChunkSize: Math.floor(total / sizes.length),
    totalCharacters: total,
  };
}

/**
 * Options for creating metadata.
 */
export type CreateMetadataOptions = {
  provider: string;
  model: string;
  dimensions: number;
  chunkSize: number;
  useMarkdownChunking: boolean;
  includeHeaders: boolean;
  headerWeight: number;
  documents: Array<{text: string; metadata?: Record<string, unknown>}>;
  chunks: Array<{text: string}>;
};

/**
 * Create metadata object for the embedding database.
 */
export function createMetadata(
  options: CreateMetadataOptions,
): EmbeddingMetadata {
  const chunkStats = calculateChunkStats(options.chunks);

  const uniqueFiles = new Set<string>();
  for (const doc of options.documents) {
    const filePath = doc.metadata?.file_path || doc.metadata?.fileName;
    if (filePath) uniqueFiles.add(String(filePath));
  }

  return {
    version: '1.0',
    createdAt: new Date().toISOString(),
    embedding: {
      provider: options.provider,
      model: options.model,
      dimensions: options.dimensions,
    },
    chunking: {
      strategy: options.useMarkdownChunking ? 'markdown-aware' : 'size-based',
      chunkSize: options.chunkSize,
      includeHeaders: options.includeHeaders,
      headerWeight: options.includeHeaders ? options.headerWeight : 0,
    },
    sourceDocuments: {
      totalDocuments: options.documents.length,
      uniqueFiles: uniqueFiles.size,
      totalCharacters: options.documents.reduce(
        (sum, d) => sum + d.text.length,
        0,
      ),
    },
    chunks: chunkStats,
    capabilities: {
      hybridSearch: true,
      ftsEnabled: true,
      sourceDocumentsStored: true,
    },
  };
}

/**
 * Store metadata in the database.
 */
export async function storeMetadataInDb(
  connector: DuckDbConnector,
  databaseName: string,
  metadata: EmbeddingMetadata,
): Promise<void> {
  const flatMetadata: Record<string, string> = {
    version: metadata.version,
    created_at: metadata.createdAt,
    embedding_provider: metadata.embedding.provider,
    embedding_model: metadata.embedding.model,
    embedding_dimensions: String(metadata.embedding.dimensions),
    chunking_strategy: metadata.chunking.strategy,
    chunk_size: String(metadata.chunking.chunkSize),
    include_headers: String(metadata.chunking.includeHeaders),
    header_weight: String(metadata.chunking.headerWeight),
    total_source_documents: String(metadata.sourceDocuments.totalDocuments),
    unique_files: String(metadata.sourceDocuments.uniqueFiles),
    source_total_characters: String(metadata.sourceDocuments.totalCharacters),
    total_chunks: String(metadata.chunks.totalChunks),
    min_chunk_size: String(metadata.chunks.minChunkSize),
    max_chunk_size: String(metadata.chunks.maxChunkSize),
    median_chunk_size: String(metadata.chunks.medianChunkSize),
    mean_chunk_size: String(metadata.chunks.meanChunkSize),
    chunks_total_characters: String(metadata.chunks.totalCharacters),
    hybrid_search_enabled: String(metadata.capabilities.hybridSearch),
    fts_enabled: String(metadata.capabilities.ftsEnabled),
    source_documents_stored: String(
      metadata.capabilities.sourceDocumentsStored,
    ),
  };

  for (const [key, value] of Object.entries(flatMetadata)) {
    await connector.query(`
      INSERT INTO ${databaseName}.embedding_metadata (key, value)
      VALUES ('${key}', '${value.replace(/'/g, "''")}')
      ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value
    `);
  }
}

/**
 * Update metadata when documents are added.
 */
export async function updateMetadataStats(
  connector: DuckDbConnector,
  databaseName: string,
  newChunks: Array<{text: string}>,
  newDocuments: Array<{text: string}>,
): Promise<void> {
  // Get current stats
  const currentStats = await connector.query(`
    SELECT key, value FROM ${databaseName}.embedding_metadata
    WHERE key IN ('total_chunks', 'total_source_documents', 'chunks_total_characters', 'source_total_characters')
  `);

  const statsMap = new Map<string, number>();
  for (const row of currentStats.toArray()) {
    statsMap.set(row.key as string, parseInt(row.value as string, 10) || 0);
  }

  const newChunkStats = calculateChunkStats(newChunks);
  const newDocChars = newDocuments.reduce((sum, d) => sum + d.text.length, 0);

  const updates: Record<string, string> = {
    total_chunks: String(
      (statsMap.get('total_chunks') || 0) + newChunkStats.totalChunks,
    ),
    total_source_documents: String(
      (statsMap.get('total_source_documents') || 0) + newDocuments.length,
    ),
    chunks_total_characters: String(
      (statsMap.get('chunks_total_characters') || 0) +
        newChunkStats.totalCharacters,
    ),
    source_total_characters: String(
      (statsMap.get('source_total_characters') || 0) + newDocChars,
    ),
  };

  for (const [key, value] of Object.entries(updates)) {
    await connector.query(`
      INSERT INTO ${databaseName}.embedding_metadata (key, value)
      VALUES ('${key}', '${value}')
      ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value
    `);
  }
}
